### ISSUE TITLE
Command line string => process.argv => how?

### ISSUE THREAD

**Issue by ORESoftware** [nodejs/help#1218]
I am looking for an answer to this question:
https://www.quora.com/unanswered/What-piece-of-code-does-Node-js-use-to-parse-command-line-argument-string-into-process-argv

the question is:
What piece of code does Node.js use to parse command line argument string into process.argv?

I'd like to use that code myself if I can. I need to parse an arbitrary (command) string into an array, and that pass that array into a routine.

**Comment by ryzokuken** [comment]
@ORESoftware the chain starts with the actual `main` function inside of `node_main.cc` which calls the `node::Start` function present inside `node.cc`which calls `env.Start`. `Environment::Start` in `env.cc` calls `SetupProcessObject` which is present in `node_internals.h`.

It's here than the magic happens:

```cc
// process.argv
  Local<Array> arguments = Array::New(env->isolate(), argc);
  for (int i = 0; i < argc; ++i) {
    arguments->Set(i, String::NewFromUtf8(env->isolate(), argv[i]));
  }
  process->Set(FIXED_ONE_BYTE_STRING(env->isolate(), "argv"), arguments);
```

in short, it seems really complicated. Please don't try this at home, if possible.

**Comment by richardlau** [comment]
>  What piece of code does Node.js use to parse command line argument string into `process.argv`?
>
> I'd like to use that code myself if I can. I need to parse an arbitrary (command) string into an array, and that pass that array into a routine.

Short answer is there isn't any parsing code as the command line arguments are already parsed into an array before Node.js starts. Node.js is a command line C/C++ program so is automatically given the number of arguments in `argc` and an array of pointers to each argument `argv` to its `main` function. This is passed around but ultimately ends up in the code quoted by @ryzokuken where JavaScript arrays and strings are created and set on the `process` object.

**Comment by advanceddeveloper** [comment]
@ORESoftware 

> What piece of code does Node.js use to parse command line argument string into process.argv

It is actually a task of the operating system. When you call `node app.js` (for example from command line), then the command line tool parses the arguements and send them to operating system, which then calls Node.js.

> I need to parse an arbitrary (command) string into an array, and that pass that array into a routine.

How command line string are parsed depends on command line tool and operating system. As I know, it is not something which has been standardized.

**Comment by ORESoftware** [comment]
@advanceddeveloper @richardlau 

thanks, what you both said makes sense...maybe it is the OS

so my use case is this:

myprogram --arg1=foo --arg2="--a=bar --z=baz"

I want myprogram to be able to pass arguments to another executable.

So my myprogram can successfully parse things so that

arg2: "--a=bar --z=baz"

so I have a naked string, but I need to parse that too. The *only* thing I can think of, is to send that string to a child process, so that it can parse it and then pass back an array through some form of IPC, maybe just stdout.

**Comment by ryzokuken** [comment]
@ORESoftware for parsing optstrings, use yargs (http://yargs.js.org/)

**Comment by ORESoftware** [comment]
@ryzokuken thanks, I have heard of yargs, let me check it out.. I created this:
https://github.com/ORESoftware/parse-command

it just uses a child process/bash to interpret a string as a command.
I simply send `process.argv` from child back to parent.

**Comment by advanceddeveloper** [comment]
@ORESoftware 

Converting a string to argument list is something which can be done by just a few lines of JavaScript code, or am I missing something? What difficulties are you having to split the string at white spaces and then parse all elements recursively, until you ent up with an array of arguments?

**Comment by the1mills** [comment]
Nah it's harder than you think when there is whitespace and quote characters

```js
node . --foo="-x -z b3"
```

Now we want to parse foo, inside our program, that's easy. But what about now:

```js
node . --foo="-x -z -a="dog cat bear" --far='oot n aboot' "
```

What i am doing is passing arguments to be interpeted by a different process than the current one.

**Comment by advanceddeveloper** [comment]
First off, it is important to precisely define how your nested quote sequences should be parsed. For example, windows command prompt parses it as `["node", ".", "--foo=-x -z -a=dog", "cat", "bear --far='oot n aboot' "]` which is, I suppose, not what you're expecting to see. Therefore, according to the fact that your needs don't match one of the implemented parsers (command prompt), I think you should define how the output should look line and based on which criterion the string is parsed in one way and not in another.

While there are two types of parentheses: open and closed ones, there are only one type of quotation marks. Therefore, it is, in general, impossible to detect which quotation characters are "open" and which are "closed". Just as an example, suppose you want to parse `"2" + 3 "-4"` where quotation marks represent parentheses. Clearly, it is impossible to uniquely parse it, because there are two valid results:

![1](https://user-images.githubusercontent.com/38354752/38946425-b55ee55e-4339-11e8-97f7-65122c4c4e7a.png)

Both are correct and unless you provide some criterion to reject one of the possible parsed results, they are still valid. Similarly, your example string could be parsed like this:

```console
["node", ".", "--foo=-x -z -a=dog", "cat", "bear --far='oot n aboot' "]
```

Is there any programmable criterion which will reject the above array as a valid result, except the "it doesn't do what I want" explanation? If yes, let me know and I would help you implement it.

**Comment by ORESoftware** [comment]
Well to be perfectly straightforward, Windows does parse commands differently. I generally don't bother to support Windows with whatever software I am writing these days.

My lib is simply designed to parse the command line args in the exact way bash would, since bash ends up doing it for me, I use the bash routine, by using a child process. (At least I think that's what is going on).

Can you try using GitBash terminal on Windows and see if the arguments are passed like Bash would do on MacOS/*nix?

**Comment by advanceddeveloper** [comment]
Here is the bash langauge specification: [Shell Command Language](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html). It isn't something which can't be implemented in JavaScript. But, of course, it requires a lot of time to implement it. I hope you don't expect from me to spend few days just to implement it for you, but I can give you some advices on how to implement some parts of it which, I think, might be problematic.

For example, parsing nested quotation marks can be done using [Precedence climbing algorithm](https://en.wikipedia.org/wiki/Operator-precedence_parser) or [Shunting-yard algorithm](https://en.wikipedia.org/wiki/Shunting-yard_algorithm). They are used for parsing mathematical expressions, but they work pretty well with any kind of nested objects. I implemented it in python few years ago and it wasn't too hard.

**Comment by ORESoftware** [comment]
@advanceddeveloper yeah I think it would be fun to try to implement it, but for now I think using a child process to interpret for you is more reliable (100% reliable by definition)... (but simply less performant).

**Comment by advanceddeveloper** [comment]
@ORESoftware This issue has been inactive for more than 8 days, so I think it would be a good idea to close it if you don't have any additional question.
